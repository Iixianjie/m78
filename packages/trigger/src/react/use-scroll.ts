import { RefObject, useEffect, useMemo, useRef } from "react";

import { dumpFn } from "@m78/utils";
import {
  createScrollTrigger,
  type ScrollTriggerInstance,
  type ScrollTriggerOption,
  type ScrollTriggerState,
} from "../scroll/scroll.js";
import { getRefDomOrDom, useFn } from "@m78/hooks";
import { isArray } from "lodash";

/** 用于use-scroll的创建配置 */
export interface UseScrollTriggerOption
  extends Omit<ScrollTriggerOption, "target" | "handle"> {
  /** 传入要绑定的滚动元素或ref, 也可以通过useScroll返回的instance.ref绑定到dom */
  el?: HTMLElement | RefObject<any>;
  /** 滚动时触发 */
  onScroll?(meta: ScrollTriggerState): void;
}

/** use-scroll扩展后的ScrollTrigger实例 */
export interface UseScrollTriggerInstance<ElType extends HTMLElement>
  extends ScrollTriggerInstance {
  /** 可使用此项代替option.el进行绑定 */
  ref: RefObject<ElType>;
}

export function useScroll<ElType extends HTMLElement>(
  option: UseScrollTriggerOption
) {
  // 用于返回的节点获取ref
  const ref = useRef<ElType>(null);

  const instance = useMemo(() => {
    return {
      ref,
      scroll: dumpFn,
      scrollToE: dumpFn,
      scrollToElement: dumpFn,
      get: dumpFn,
      destroy: dumpFn,
    } as any as UseScrollTriggerInstance<ElType>;
  }, []);

  const handle = useFn((e: ScrollTriggerState) => {
    option.onScroll?.(e);
  });

  // 在满足条件时重载实例
  const getDeps = () => {
    const _offset = option.offset;
    const offset = isArray(_offset) ? _offset : [_offset, _offset];

    const _touchOffset = option.touchOffset;
    const touchOffset = isArray(_touchOffset)
      ? _touchOffset
      : [_touchOffset, _touchOffset];

    return [
      option.el,
      ref.current,
      (option.el as any)?.current,
      offset[0],
      offset[1],
      touchOffset[0],
      touchOffset[1],
    ];
  };

  useEffect(() => {
    const curEl = getRefDomOrDom(option.el, ref);

    if (curEl) {
      Object.assign(
        instance,
        createScrollTrigger({
          ...option,
          target: curEl,
          handle,
        })
      );
    }

    return () => {
      instance.destroy?.();
    };
  }, getDeps());

  return instance;
}
