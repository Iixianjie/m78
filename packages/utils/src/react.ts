import { AnyFunction } from "./types";

/**
 * 便捷的按键和点击事件绑定
 * @param handle - 时间处理函数
 * @param spaceTrigger - 按下空格时是否触发
 * */
export function keypressAndClick(handle: AnyFunction, spaceTrigger = true) {
  return {
    onClick: handle,
    onKeyPress: (e: any) => {
      const code = e.code;

      if (code === "Enter" || (spaceTrigger && code === "Space")) {
        handle?.();
      }
    },
  };
}
