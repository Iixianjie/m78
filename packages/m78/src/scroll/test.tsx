import { render } from "@testing-library/react";
import React from "react";
import { Scroll } from "./index";

describe("Scroll", () => {
  test("base", () => {
    const r = render(
      <Scroll
        direction="xy"
        style={{ height: 300, width: 300, border: "1px solid red" }}
        dragScroll
        scrollIndicator={false}
      >
        <div
          style={{
            width: 600,
            height: 600,
          }}
        >
          lorem
        </div>
      </Scroll>
    );

    expect(r.container).toMatchSnapshot();
  });
});
