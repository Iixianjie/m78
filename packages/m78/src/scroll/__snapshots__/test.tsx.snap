// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll base 1`] = `
<div>
  <div
    class="m78 m78-scroll __xy"
    style="height: 300px; width: 300px; border: 1px solid red;"
  >
    <div
      class="m78-scroll_wrap"
      style="overflow-x: scroll; overflow-y: scroll; bottom: -40px; right: -40px; transform: none; user-select: none;"
    >
      <div
        class="m78-scroll_cont"
      >
        <div
          style="width: 600px; height: 600px;"
        >
          lorem
        </div>
      </div>
    </div>
    <div
      class="m78-scroll_bar __x"
      style="display: none;"
    >
      <div
        class="m78-scroll_bar_thumb"
        style="width: 0px; transform: none;"
      />
    </div>
    <div
      class="m78-scroll_bar __y"
      style="display: none;"
    >
      <div
        class="m78-scroll_bar_thumb"
        style="height: 0px; transform: none;"
      />
    </div>
  </div>
</div>
`;
